pkg/ebpf/c/tracee.bpf.c

```diff
@@ -1741,146 +1751,161 @@ static __always_inline int init_event_data(event_data_t *data, void *ctx)
 
 // INTERNAL: FILTERING -----------------------------------------------------------------------------
 
-static __always_inline int
-uint_filter_matches(bool filter_out, void *filter_map, u64 value, u64 max, u64 min)
+static __always_inline u32
+uint_filter_matches(u32 filter_out, void *filter_map, u64 value, u64 max, u64 min)
 {
-    u8 *equality = bpf_map_lookup_elem(filter_map, &value);
-    if (equality != NULL)
-        return *equality;
+    u32 equality_bitmask = 0;
+    u32 equality_valid_bits = 0;
+    eq_t *equality = bpf_map_lookup_elem(filter_map, &value);
+    if (equality != NULL) {
+        equality_bitmask = equality->bitmask;
+        equality_valid_bits = equality->valid_bits;
+    }
 
+    // we can keep this logic by having one global lessThan and greaterThan per each context filter type
+    // if there is more than one filter used, it can be checked in userspace. (very unlikely).
+    // there will always be a valid (scope) bitmap in addition to the returned value.
     if ((max != FILTER_MAX_NOT_SET) && (value >= max))
-        return 0;
+        return equality_bitmask;
 
     if ((min != FILTER_MIN_NOT_SET) && (value <= min))
-        return 0; // 0 means do not trace
+        return equality_bitmask;
```

Comment:
@yanivagman I made some tests and this logic seems not work for multiple scopes. When we have, for instance, two scopes like:

```
-t "comm=ls pid>256000 pid<256900" \
-t "comm=uname uts=hb"
```

the uname comm is not submitted when not in the min max range (since it's global).

I suppose that all min max filtering, for multiple scope, needs to take place in user space only.

The behaviour must be discussed as well for cases like:

```
-t "comm=ls pid!=1" \
-t "pid=1"
```

The key is the same, but multiple scopes require events to be submitted for both cases.

Reply:
After discussion with @yanivagman I fixed and implemented lacking logic. 👍🏼